### PR TITLE
feat (algo): Add MAP-Elites Low-Spread

### LINKED ISSUE #151: Add MAP-Elites Low-Spread
Implement ME-LS from the QD Transformer paper: https://dl.acm.org/doi/abs/10.1145/3583131.3590433

### PR DESCRIPTION
**Pull request by btjanaka**
Related issues: [refer to issues] Resolves #151 

[Introduce the overall change made in the PR and list the associated modifications below]

Add MAP-Elites Low-Spread: https://dl.acm.org/doi/abs/10.1145/3583131.3590433

This PR introduces:
- `MELSRepertoire` in `mels_repertoire.py`
  - Tests in `mels_repertoire_test.py`
- `MELS` in `mels.py`
  - Tests in `mels_test.py`
- `Spread` type in `types.py`
- `multi_sample_scoring_function` in `sampling.py`

## Questions

1. Should we store all `n_evals` fitnesses and descriptors in the repertoire, or should we just store the aggregated fitness and descriptors? We can also store both the aggregate and the separate values. Storing separate values would require passing in `n_evals` to `MELSRepertoire.init_default` so that we can define the shapes of `fitnesses` and `descriptors`.
2. Should we check if one individual dominates the others if inserting multiple individuals into the same cell?
3. What link should I use for citations? Currently I am using the ACM Digital Library link. I can also use the arXiv link: https://arxiv.org/abs/2303.16207 if preferred.

## Checks

- [x] a clear description of the PR has been added
- [x] sufficient tests have been written
- [x] relevant section added to the documentation
- [x] example notebook added to the repo
- [x] clean docstrings and comments have been written
- [x] if any issue/observation has been discovered, a new issue has been opened

## Future improvements

[List here potential observations made and/or improvements that could be made in the future. If relevant, open issues for those.]

### CODE REVIEW

**Comment by Lookatator** [comment]
Hi @btjanaka ! Thank you very much for this (super well-documented) PR.

I added @Egiob and @manon-but-yes as reviewers for this PR; they are probably the best people to judge it.

I had an overview of the code and it looks great! I have mainly one high-level comment regarding the implementation of the reevaluation scoring function (and where it should be implemented). I will add it as a comment.

File: qdax/core/containers/mels_repertoire.py [hunk 1]
```diff
@@ -0,0 +1,334 @@
...
+
+    # each row in xy needs to be added with x2[i]
+    # each column of xy needs to be added with y2[j]
+    # to get everything to play well, we'll need to reshape
+    # x2 from (m) -> (m, 1), numpy will handle the rest of the broadcasting for us
+    # see: https://numpy.org/doc/stable/user/basics.broadcasting.html
+    x2 = x2.reshape(-1, 1)
+    dists = jnp.sqrt(
+        x2 - 2 * xy + y2
+    )  # (m, 1) repeat columnwise + (m, n) + (n) repeat rowwise -> (m, n)
```
[the remark this answers is not in the document]
Comment (btjanaka) [reply qdax/core/containers/mels_repertoire.py:54]: Good point; I believe the one-liner is `jnp.linalg.norm(x[:, None] - y)`. I'll update accordingly.

[the remark this answers is not in the document]
Comment (btjanaka) [reply qdax/core/containers/mels_repertoire.py]: *actually, it's `jnp.linalg.norm(x[:, None] - y, axis=2)`

**Comment by btjanaka** [comment]
Hi @Lookatator, thank you for taking a look at this PR! I addressed your comment on the pairwise distance calculation, but I do not see the comment on scoring function evaluation. Did you make the comment in the tests or the notebook? (the scoring function is in both those places).

**Comment by Egiob** [comment]
Hi guys! I will do a pass by the end of the week :smile:

File: tests/core_test/mels_test.py [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+        extra_scores["transitions"] = jax.tree_util.tree_map(
+            lambda x: x.reshape((batch_size, num_evals, *x.shape[1:])),
+            extra_scores["transitions"],
+        )
+        return (
+            fitnesses.reshape((batch_size, num_evals)),
+            descriptors.reshape((batch_size, num_evals, -1)),
+            extra_scores,
+            random_key,
+        )
```
Comment (Lookatator) [inline tests/core_test/mels_test.py:136]: I discussed the way the scoring function is defined here with @limbryan,  and we think that `scoring_fn` should be as general as possible. For example, we should define a `scoring_fn` that could be passed to `MELS` but also to the standard `MAPElites` class.

To do that, I suggest that you:
- pass the standard `scoring_fn` to the `MELS` class.
- and in the `MELS` class, you override `update(...)` with a code that does a `vmap` on the scoring fn to evaluate it on several individuals. Here is an approximate suggestion on how to change `scoring_function` call in the `update(...)` method.
```python
...
random_key, *subkeys = jax.random.split(random_key, self.num_evals + 1)  # num_evals can be an attribute of the `MELS` class
fitnesses, descriptors, extra_scores, _ = jax.vmap(self.scoring_function, in_axes=(None, 0))(genotypes, jnp.asarray(subkeys))
...  #  adding to container and so on...
```

Comment (btjanaka) [reply tests/core_test/mels_test.py:136]: I think it makes sense to have a scoring_fn that is the same as in MAPElites, but one issue I see is that the brax scoring function takes in a set of `init_states`. Using the vmap approach, it seems the `init_states` would be repeated across all the evals, thus resulting in 0 spread (IIRC the initial state is the only non-deterministic component in these environments). 

One potential idea (albeit a backwards-incompatible one) is that perhaps `scoring_function_brax_envs` should actually be taking in a series of random keys, one for each controller it is evaluating. That way, if you want to create different environments, you can pass in a ton of random keys, and if you don't want stochasticity, you pass in the same random key repeated several times.

Comment (limbryan) [reply tests/core_test/mels_test.py:136]: Not off the top of my head (the tutorials need some updating in general but haven't had the time to get around it) but ```env.reset``` or ```partial(env.reset)``` should work as the ```play_reset_fn```

Comment (btjanaka) [reply tests/core_test/mels_test.py]: @Lookatator @limbryan I finished implementing this change! I added a `multi_eval_scoring_function` in `MELS.__init__`. The scoring function now uses `reset_based_scoring_function_brax_envs`

File: qdax/core/mels.py [hunk 3]
```diff
@@ -0,0 +1,120 @@
...
+            random_key = subkeys[0]
+            fitnesses, descriptors, extra_scores, _ = jax.vmap(
+                scoring_function, in_axes=(None, 0)
+            )(genotypes, subkeys[1:])
+
+            # The vmap will output arrays with shape (num_evals, batch_size, ...)
+            # but we need to have (batch_size, num_evals, ...) for MELSRepertoire.
+            fitnesses = jax.tree_map(lambda x: jnp.swapaxes(x, 0, 1), fitnesses)
+            descriptors = jax.tree_map(lambda x: jnp.swapaxes(x, 0, 1), descriptors)
+            extra_scores = jax.tree_map(lambda x: jnp.swapaxes(x, 0, 1), extra_scores)
```
Comment (Lookatator) [inline qdax/core/mels.py:60]: [Optional Tip] To avoid having to do this, I think you can simply add the argument `out_axes=1` to the `vmap` above.

File: tests/core_test/mels_test.py (hunk 2 above)
Comment (Lookatator) [reply tests/core_test/mels_test.py]: Great! Thank you @btjanaka!

File: tests/core_test/containers_test/mels_repertoire_test.py [hunk 4]
```diff
@@ -0,0 +1,230 @@
+import jax.numpy as jnp
+import pytest
+
+from qdax.core.containers.mels_repertoire import MELSRepertoire
+from qdax.types import ExtraScores
+
+
+def test_mels_repertoire() -> None:
```
Comment (Egiob) [inline tests/core_test/containers_test/mels_repertoire_test.py:8]: Minor suggestion: we could use a short sentence to define the scope of the test

File: tests/core_test/containers_test/mels_repertoire_test.py [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+        pytest.assume(
+            jnp.allclose(repertoire.fitnesses, expected_fitnesses_2, atol=1e-6)
+        )
+        pytest.assume(
+            jnp.allclose(repertoire.descriptors, expected_descriptors_2, atol=1e-6)
+        )
+        pytest.assume(jnp.allclose(repertoire.spreads, expected_spreads_2, atol=1e-6))
+
+
+def test_single_eval() -> None:
```
Comment (Egiob) [inline tests/core_test/containers_test/mels_repertoire_test.py:170]: Minor suggestion: we could use a short sentence to define the scope of the test

File: qdax/core/containers/mels_repertoire.py [hunk 6]
```diff
@@ -0,0 +1,311 @@
...
+from jax.flatten_util import ravel_pytree
+
+from qdax.core.containers.mapelites_repertoire import (
+    MapElitesRepertoire,
+    get_cells_indices,
+)
+from qdax.types import Centroid, Descriptor, ExtraScores, Fitness, Genotype, Spread
+
+
+def _dispersion(descriptors: jnp.ndarray) -> jnp.ndarray:
```
Comment (Egiob) [inline qdax/core/containers/mels_repertoire.py:20]: Minor suggestion: perhaps these functions belong elsewhere in a utils/ file ? I feel like they can be used by other methods as scores or metrics

File: qdax/core/mels.py [hunk 7]
```diff
@@ -0,0 +1,120 @@
...
+    def __init__(
+        self,
+        scoring_function: Callable[
+            [Genotype, RNGKey], Tuple[Fitness, Descriptor, ExtraScores, RNGKey]
+        ],
+        emitter: Emitter,
+        metrics_function: Callable[[MELSRepertoire], Metrics],
+        num_evals: int,
+    ) -> None:
+        def multi_eval_scoring_function(
```
Comment (Egiob) [inline qdax/core/mels.py:46]: Minor suggestion: put this function on top of the file with a `make_multi_eval_scoring_function` to reduce nested indents.

**Review by Egiob — APPROVED** [review]
Hey @btjanaka, thanks for the very-clean PR, amazing work! I've left a couple of small suggestions, but all good for me :+1:

File: qdax/core/mels.py (hunk 7 above)
Comment (btjanaka) [reply qdax/core/mels.py:46]: Agreed

File: qdax/core/containers/mels_repertoire.py (hunk 6 above)
Comment (btjanaka) [reply qdax/core/containers/mels_repertoire.py:20]: I think if it becomes something used in multiple places, perhaps it can then be refactored? Since it is a private function, it seems it can be moved around pretty easily.

File: tests/core_test/containers_test/mels_repertoire_test.py (hunk 5 above)
Comment (btjanaka) [reply tests/core_test/containers_test/mels_repertoire_test.py]: Updated the test names and added docstrings

File: tests/core_test/containers_test/mels_repertoire_test.py (hunk 4 above)
Comment (btjanaka) [reply tests/core_test/containers_test/mels_repertoire_test.py]: Updated the test names and added docstrings

**Comment by btjanaka** [comment]
Thank you @Egiob! I've resolved all your comments except for the one about moving `_dispersion`

**Comment by manon-but-yes** [comment]
Hi @btjanaka thank you for this PR! Apologies for the delay I just got back to work.

File: qdax/core/mels.py [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+    Descriptor,
+    ExtraScores,
+    Fitness,
+    Genotype,
+    Metrics,
+    RNGKey,
+)
+
+
+def multi_eval_scoring_function(
```
Comment (manon-but-yes) [inline qdax/core/mels.py:23]: There is a file containing a similar function in qdax/utils/sampling.py, that is uses by ME-ES, maybe we want to uniform this? 
multi_eval_scoring_function() seems to be a subcase of sampling() that returns all the evaluations, while sampling returns the final average, so sampling() could just wrap multi_eval_scoring_function().

Comment (btjanaka) [reply qdax/core/mels.py]: Hi @manon-but-yes, thank you for the feedback! I have created a `multi_sample_scoring_function` in `sampling.py` and removed `multi_eval_scoring_function` from `mels.py`. I also changed `num_evals` to `num_samples` everywhere I could find it. Let me know what you think!

Comment (manon-but-yes) [reply qdax/core/mels.py]: It's perfect!

**Review by manon-but-yes — APPROVED** [review]

**Review by Lookatator — APPROVED** [review]
